common/lib/plugins/failover2/failover2_plugin.ts

```diff
@@ -0,0 +1,502 @@
+
+  private hostListProviderService?: HostListProviderService;
+  protected enableFailoverSetting: boolean = WrapperProperties.ENABLE_CLUSTER_AWARE_FAILOVER.defaultValue;
+  private failoverTimeoutSettingMs: number = WrapperProperties.FAILOVER_TIMEOUT_MS.defaultValue;
+
+  constructor(pluginService: PluginService, properties: Map<string, any>, rdsHelper: RdsUtils) {
+    super();
+    this._properties = properties;
+    this.pluginService = pluginService;
+    this._rdsHelper = rdsHelper;
+    this._staleDnsHelper = new StaleDnsHelper(this.pluginService);
+    this.enableFailoverSetting = WrapperProperties.ENABLE_CLUSTER_AWARE_FAILOVER.get(this._properties);
+    this.failoverTimeoutSettingMs = WrapperProperties.FAILOVER_TIMEOUT_MS.get(this._properties);
+
+    const telemetryFactory = this.pluginService.getTelemetryFactory();
+    this.failoverWriterTriggeredCounter = telemetryFactory.createCounter("writerFailover.triggered.count");
+    this.failoverWriterSuccessCounter = telemetryFactory.createCounter("writerFailover.completed.success.count");
+    this.failoverWriterFailedCounter = telemetryFactory.createCounter("writerFailover.completed.failed.count");
+    this.failoverReaderTriggeredCounter = telemetryFactory.createCounter("readerFailover.triggered.count");
+    this.failoverReaderSuccessCounter = telemetryFactory.createCounter("readerFailover.completed.success.count");
+    this.failoverReaderFailedCounter = telemetryFactory.createCounter("readerFailover.completed.failed.count");
+  }
+
+  override getSubscribedMethods(): Set<string> {
+    return Failover2Plugin.subscribedMethods;
+  }
+
+  override initHostProvider(
+    hostInfo: HostInfo,
+    props: Map<string, any>,
+    hostListProviderService: HostListProviderService,
+    initHostProviderFunc: () => void
+  ): void {
+    this.hostListProviderService = hostListProviderService;
+    if (!this.enableFailoverSetting) {
+      return;
+    }
+
+    initHostProviderFunc();
+
+    this.failoverMode = failoverModeFromValue(WrapperProperties.FAILOVER_MODE.get(this._properties));
+    this._rdsUrlType = this._rdsHelper.identifyRdsType(hostInfo.host);
+
+    if (this.failoverMode === FailoverMode.UNKNOWN) {
+      this.failoverMode = this._rdsUrlType === RdsUrlType.RDS_READER_CLUSTER ? FailoverMode.READER_OR_WRITER : FailoverMode.STRICT_WRITER;
+    }
+
+    logger.debug(Messages.get("Failover.parameterValue", "failoverMode", FailoverMode[this.failoverMode]));
+  }
+
+  override notifyConnectionChanged(changes: Set<HostChangeOptions>): Promise<OldConnectionSuggestionAction> {
+    return Promise.resolve(OldConnectionSuggestionAction.NO_OPINION);
+  }
+
+  private isFailoverEnabled(): boolean {
+    return (
+      this.enableFailoverSetting &&
+      this._rdsUrlType !== RdsUrlType.RDS_PROXY &&
+      this.pluginService.getHosts() &&
+      this.pluginService.getHosts().length > 0
+    );
+  }
+
+  private async invalidInvocationOnClosedConnection() {
+    if (!this._closedExplicitly) {
```

Comment:
Is `this._closedExplicitly` ever updated? I only see it being initialized to `false`.